src/mito2/src/wal/entry_distributor.rs

```diff
@@ -0,0 +1,643 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::min;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use api::v1::WalEntry;
+use async_stream::stream;
+use common_telemetry::{debug, error};
+use futures::future::join_all;
+use snafu::ensure;
+use store_api::logstore::entry::Entry;
+use store_api::logstore::provider::Provider;
+use store_api::storage::RegionId;
+use tokio::sync::mpsc::{self, Receiver, Sender, UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot;
+use tokio_stream::StreamExt;
+
+use crate::error::{self, Result};
+use crate::wal::entry_reader::{decode_raw_entry, WalEntryReader};
+use crate::wal::raw_entry_reader::RawEntryReader;
+use crate::wal::{EntryId, WalEntryStream};
+
+/// [WalEntryDistributor] distributes Wal entries to specific [WalEntryReceiver]s based on [RegionId].
+pub(crate) struct WalEntryDistributor {
+    raw_wal_reader: Arc<dyn RawEntryReader>,
+    provider: Provider,
+    /// Sends [Entry] to receivers based on [RegionId]
+    senders: HashMap<RegionId, Sender<Entry>>,
+    /// Waits for the arg from the [WalEntryReader].
+    arg_receivers: Vec<(RegionId, oneshot::Receiver<EntryId>)>,
+}
+
+impl WalEntryDistributor {
+    /// Distributes entries to specific [WalEntryReceiver]s based on [RegionId].
+    pub async fn distribute(mut self) -> Result<()> {
+        let args_futures = self
+            .arg_receivers
+            .iter_mut()
+            .map(|(region_id, receiver)| async {
+                {
+                    if let Ok(start_id) = receiver.await {
+                        (*region_id, Some(start_id))
+                    } else {
+                        (*region_id, None)
+                    }
+                }
+            });
+        let args = join_all(args_futures)
+            .await
+            .into_iter()
+            .filter_map(|(region_id, start_id)| start_id.map(|start_id| (region_id, start_id)))
+            .collect::<Vec<_>>();
+
+        // No subscribers
+        if args.is_empty() {
+            return Ok(());
+        }
+        // Safety: must exist
+        let mut min_start_id = args.first().unwrap().1;
+
+        let mut subscribers = HashMap::with_capacity(args.len());
+        for (region_id, start_id) in args {
+            subscribers.insert(
+                region_id,
+                EntryReceiver {
+                    start_id,
+                    sender: self.senders[&region_id].clone(),
+                },
+            );
+            min_start_id = min(min_start_id, start_id);
+        }
```

Comment:
Maybe we using two separate loops to obtain `min_start_id` and `subscribers` is better, as two loops is more clearly, and maybe the CPU handles simple loops more efficiently.
```suggestion
        let min_start_id = args
            .iter()
            .map(|(_, start_id)| *start_id)
            .min()
            .unwrap_or_default();

        let subscribers: HashMap<_, _> = args
            .into_iter()
            .map(|(region_id, start_id)| {
                (
                    region_id,
                    EntryReceiver {
                        start_id,
                        sender: self.senders[&region_id].clone(),
                    },
                )
            })
            .collect();
```